### PR TITLE
Uplus v2

### PR DESCRIPTION
**Pull request by mtaillefumier**
This PR includes all necessary bit of code to run simulations with the LDA+U+V correction and more. It includes so far

- Multi-orbital case for both U and V corrections
- Possibility to orthogonalize the hubbard subspace compared to all atomic wave functions
- Forces (for U+V) and stresses for the co-linear case
- 3 more tests for testing the code

Remains to be done

- [ ] Move symmetrization of the occupations matrices to the right place (the function would need too many parameters, so need a proper fix for this)
- [x] Add a function for adding atom pairs for the V term in the api
- [x] Extra work on QE side
- [ ] And of course more testing
- [x] remove duplicated code for finding the orbital index

The multi-orbital case is treated such that the code keep a list of "atomic" levels to be considered in the U (and / or V) term. So instead of having occupations numbers per atoms we have occupations numbers per atomic levels. This would be equivalent to having a full matrix that we decompose per block, each block representing a term in the correction.

### CODE REVIEW

File: src/api/sirius_api.cpp [hunk 1]
```diff
@@ -1957,17 +1957,24 @@ sirius_add_atom_type_radial_function(void* const* handler__, char const* atom_ty
       type: double
       attr: in, required
       doc: J0 for the simple interaction treatment.
+      J0:
+      type: bool
+      attr: in, optional
+      doc: J0 for the simple interaction treatment.
```
Comment (toxa81) [inline src/api/sirius_api.cpp:1963]: copy-paste is wrong

File: src/hubbard/hubbard_matrix.cpp [hunk 2]
```diff
@@ -32,31 +32,59 @@ Hubbard_matrix::Hubbard_matrix(Simulation_context& ctx__)
...
+        /* the offsets here match the offsets of the hubbard wave functions but
+         * are more fine grained. The offsets of the hubbard wave functions are
+         * for atom while here they are for each atomic level. Since all atomic
+         * level of a given atom are next to each other, the offset of the first
+         * atomic level of a given atom has the same value than the offset
+         * giving the position of the first hubbard wave function of this
+         * atom. */
+        offset_ = std::vector<int>(num_atomic_level_, 0);
+
+        int size__ = 0;
```
Comment (toxa81) [inline src/hubbard/hubbard_matrix.cpp:61]: local variables do not have tailing underscores

File: src/mixer/mixer_functions.cpp [hunk 3]
```diff
@@ -399,47 +399,63 @@ FunctionProperties<Hubbard_matrix> hubbard_matrix_function_property()
...
-            }
+      for (size_t at_lvl = 0; at_lvl < x.local().size(); at_lvl++) {
+        for (size_t i = 0; i < x.local(at_lvl).size(); i++) {
+          y.local(at_lvl)[i] = alpha * x.local(at_lvl)[i] + y.local(at_lvl)[i];
         }
+      }
+
+      for (size_t link = 0; link < x.nonlocal().size(); link++) {
+        for (size_t i = 0; i < x.nonlocal(link).size(); i++) {
+          y.nonlocal(link)[i] *= alpha * x.nonlocal(link)[i] + y.nonlocal(link)[i];
```
Comment (toxa81) [inline src/mixer/mixer_functions.cpp:435]: why *= here instead of = ?

File: verification/test12/output_ref.json [hunk 4]
```diff
@@ -1,127 +1,278 @@
...
+            "bxc": -0.0016663868700351283,
+            "core_eval_sum": 0.0,
+            "entropy_sum": -0.00141047104512723,
+            "enuc": 0.0,
+            "eval_sum": -8.227567513814598,
+            "ewald": -107.33691365734347,
+            "exc": -24.980499738272602,
+            "kin": 43.778506377374406,
+            "scf_correction": 2.335812609999266e-08,
+            "total": -146.56055311695883,
```
Comment (toxa81) [inline verification/test12/output_ref.json:225]: change in total energy from `-147.35305830650776`

File: verification/test13/output_ref.json [hunk 5]
```diff
@@ -1,115 +1,290 @@
...
+            "bxc": -0.002710825433261687,
+            "core_eval_sum": 0.0,
+            "entropy_sum": -0.0014104712994883058,
+            "enuc": 0.0,
+            "eval_sum": 4.1830387955927835,
+            "ewald": -35.977192674166105,
+            "exc": -15.81795021311408,
+            "kin": 8.093820536934306,
+            "scf_correction": -1.4796822878082594e-08,
+            "total": -42.27968211789244,
```
Comment (toxa81) [inline verification/test13/output_ref.json:225]: change in total energy from `-43.07007257816769`

File: verification/test22/output_ref.json [hunk 6]
```diff
@@ -207,116 +223,116 @@
...
-            "veff": -141.24232201781993,
-            "vha": 245.71593486548946,
-            "vloc": -333.6713247750155,
-            "vxc": -53.28693210828464
+            "eval_sum": -24.38389471519399,
+            "ewald": -238.91986829593844,
+            "exc": -42.21059833693431,
+            "kin": 116.57128463090763,
+            "scf_correction": 4.865204800807987e-06,
+            "total": -375.4010540846649,
```
Comment (toxa81) [inline verification/test22/output_ref.json:248]: change in total energy from `-377.2823257239195`

**Review by toxa81 — CHANGES_REQUESTED** [review]

**Comment by mtaillefumier** [comment]
I had to revert two patch (the last two) because I could not find the error. things seem to behave now.

**Pull request closed without merging**